solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -246,6 +244,95 @@ public boolean canBeRemoved() {
     }
   }
 
+  /**
+   * A ConcurrentHashMap of active watcher by collection name
+   *
+   * <p>Each watcher DocCollectionWatch also contains the latest DocCollection (state) observed
+   */
+  private static class DocCollectionWatches
+      extends ConcurrentHashMap<String, StatefulCollectionWatch<DocCollectionWatcher>> {
+
+    /**
+     * Gets the DocCollection (state) of the collection which the corresponding watch last observed
+     *
+     * @param collection the collection name to get DocCollection on
+     * @return The last observed DocCollection(state). if null, that means there's no such
+     *     collection.
+     */
+    private DocCollection getDocCollection(String collection) {
+      StatefulCollectionWatch<DocCollectionWatcher> watch = get(collection);
+      return watch != null ? watch.currentState : null;
+    }
+
+    /**
+     * Gets the active collections (collections that exist) being watched
+     *
+     * @return an immutable set of active collection names
+     */
+    private Set<String> activeCollections() {
+      return this.entrySet().stream()
+          .filter(
+              (Entry<String, StatefulCollectionWatch<DocCollectionWatcher>> entry) ->
+                  entry.getValue().currentState != null)
+          .map(Entry::getKey)
+          .collect(Collectors.toUnmodifiableSet());
+    }
+
+    /**
+     * Updates the latest observed DocCollection (state) of the {@link StatefulCollectionWatch} if
+     * the collection is being watched
+     *
+     * @param collection the collection name
+     * @param newState the new DocCollection (state) observed
+     * @return whether an active watch exists for such collection
```

Comment:
Good question @HoustonPutman ! I remember trying to make my method to behave the same as `updateWatchedCollection`, that it returns `true` even if:
1. Both oldState and newState are null, this could happen after deletion when i debug
2. newState has older version than oldState, hence no updates are made

After I simplified my logic, i notice that the end result turns out to be just "whether the collection was watched" , hence i changed the javadoc to describe this exact behavior 😓 

By looking at the existing `updateWatchedCollection`, it appears to only return false if the collection is no longer watched (with exception that new state is null?)

Though I do share the concerns that some places that rely on the return values of the new method, might now seem  a bit odd ie:
```
if (collectionWatches.updateDocCollection(coll, newState)) {
          updatedCollections.add(coll);
```

I am not sure if I completely follow the original comments in the code
```
// no change to state, but we might have been triggered by the addition of a
          // state watcher, so run notifications
```
So is the return value actually flags whether we should run notifications (i assume it means running `constructState`?). If so, would it make sense to just call this flag `shouldReconstructState` or something? 🤔

Reply:
so `constructState` takes in the list of collections that should be notified. It constructs the whole state, then notifies the watchers of the "updated" collections afterwards.

> I am not sure if I completely follow the original comments in the code

I do not think that comment still applies... The only times where the return value of this method is used is in are:
- `forciblyRefreshAllClusterStateSlow()`
- `forceUpdateCollection(String collection)`
- `compareStateVersions(String coll, int version)`

None of these are adding a state watcher. And every creation of a `StateWatcher` uses `new StateWatcher(collection).refreshAndWatch()`

in `refreshAndWatch()` we find:
```java
DocCollection newState = fetchCollectionState(coll, this);
collectionWatches.updateDocCollection(coll, newState);
synchronized (getUpdateLock()) {
  constructState(Collections.singleton(coll));
}
```

Meaning `constructState` is always called on new StateWatchers anyways... So to end, I think we can remove the `=` in `>=` for that if statement. But we do need to adhere to the original functionality of the return statement, which is "has this state changed?"

Reply:
Thank you for the detailed explanations @HoustonPutman !

So it appears that `constructState` serves at least 2 purposes:
1. Construct the clusterstate again as there are changes in the watched collection states
2. Notify all the watchers as defined in `changedCollections` param

So i agree that for 1., it only makes sense to re-construct the state if indeed the doc collection has been changed

As for 2, I don't understand enough to tell whether it matters, like whether `forciblyRefreshAllClusterStateSlow` would expect all watches be notified, whether there's any change on the corresponding collection or not. But i trust your analysis!